checkpoint_service.py

```diff
@@ -0,0 +1,106 @@
+import os
+import logging
+from abc import ABC, abstractmethod
+
+# List of all objectTypes that we export / import in WM
+USER_OBJECT = "users"
+GROUP_OBJECT = "groups"
+WORKSPACE_NOTEBOOK_OBJECT = "notebooks"
+WORKSPACE_DIRECTORY_OBJECT = "directories"
+WORKSPACE_NOTEBOOK_ACL_OBJECT = "acl_notebooks"
+WORKSPACE_DIRECTORY_ACL_OBJECT = "acl_directories"
+METASTORE_TABLES = "metastore"
+
+# Migration pipeline placeholder constants
+MIGRATION_PIPELINE_ACTION_TYPE = "pipeline"
+MIGRATION_PIPELINE_OBJECT_TYPE = "tasks"
+
+# Actions
+WM_EXPORT = "export"
+WM_IMPORT = "import"
+
+class AbstractCheckpointKeySet(ABC):
+    """Abstract base class for checkpoint read and write."""
+
+    @abstractmethod
+    def write(self, key):
+        """Writes key into checkpoint file."""
+        pass
+
+    @abstractmethod
+    def contains(self, key):
+        """Checks if key exists in checkpoint"""
+        pass
+
+class CheckpointKeySet(AbstractCheckpointKeySet):
+    """Deals with checkpoint read and write."""
+
+    def __init__(self, checkpoint_file):
+        """
+        :param checkpoint_file: file to read / write object keys for checkpointing
+        """
+        self._checkpoint_file = checkpoint_file
+        self._checkpoint_file_append_fp = open(checkpoint_file, 'a+')
+        self._checkpoint_key_set = set()
+        self._restore_from_checkpoint_file()
+
+    def write(self, key):
+        """Writes key into checkpoint file. This also flushes data after write to prevent loss of checkpoint data on
+        system crash.
+
+        Note: Make sure to persist your data before calling write. There is risk of data loss if your data is not persisted
```

Comment:
Nit: could you make it more alarming? Something like !!!CAUTION!!!

Reply:
done